Split Validate methods

macisamuele opened the pull request:
This PR is related to #63 even if it is not fully addressing it yet.

Disclaimer: this is huge (and sorry in advance for that).

The concept of the PR is to have the a single point (or as limited as possible) where we do interact with `serde_json::Value::as_*` methods (or the logically equivalent pattern matching).

In order to achieve so the `Validate` trait needs to define a set of `is_valid_*` and `validate_*` methods.
By default (this is mostly done to reduce code verbosity if possible) the trait defines:
 *  `is_valid_*` as always returning true (essentially the `TrueValidator`)
 * `validate_*` methods do use the result of the matching `is_valid_*` to decide if they have to return `no_error` or a specific error (built via `Validate:: build_validation_error`)
 * `build_validation_error` to generate a `ValidationError` for the current instance (meant to reduce duplication and ensure that default implementation of `validate_*` makes sense)

Something to notice here is that I've not yet checked for benchmark results (that's one of the reason for the last commit in this PR) but I might eventually expect some small regression as all the code that has been created does not really have any inlining (not considering to do it now considering the huge size) but at the same time we should have some lifting from reduced interaction with `Value::as_*` methods

If tests are green (and they should) it would be mergeable, but considering the size and the few tricks and twists of few validators I would highly recommend a proper review.

Stranger6667 commented:
Cool! Looking at it ... Could you, please move changes in workflow.yaml & benchmark into a separate PR? so it is more convenient to run the benchmarks...

macisamuele commented:
Going to move the last commit out of this PR into #87.

Stranger6667 commented on src/validator.rs:
```diff
@@ -0,0 +1,272 @@
+    }
+    fn name(&self) -> String;
+
+    #[inline(always)]
```
I am not completely sure about `always`, the `#[inline]` version gave me ~5% improvement over big/small schemas validation in this branch

macisamuele replied:
This is actually a surprise to me, I was almost certain that I did remove all the `#[inline]` attributes as adding them just because "it feels a good idea" is generally not (except if guided by benchmarking).
I'm going to remove them and re-add them as needed after benchmarking (maybe after this landed on master)

macisamuele commented on src/keywords/enum_.rs:
```diff
@@ -45,6 +40,57 @@ impl Validate for EnumValidator {
+        })
+    }
+    fn is_valid_null(&self, _: &JSONSchema, _: &Value, _: ()) -> bool {
+        self.items.iter().any(|item| item.is_null())
```
Going to address it as well as other `clippy::pedantic` reports (I'm going to ignore the `clippy::module_name_repetitions` ones)

Stranger6667 commented:
Could you, please, rebase on master? 
Unfortunately locally it gives me ~7-12% performance regression :( Will try to take a look with `perf` tomorrow

macisamuele commented:
> Could you, please, rebase on master?

✔️ 

> Unfortunately locally it gives me ~7-12% performance regression :( Will try to take a look with `perf` tomorrow

There are good chances that removing the `#[inline]` in 9340b02 (specifically on `validator::helpers::*`) might causing this regression 🤞

Stranger6667 commented on src/keywords/items.rs:
```diff
@@ -22,103 +25,101 @@ impl ItemsArrayValidator {
-                    .collect()
-            };
-            return Box::new(errors.into_iter());
+                    .all(|item| helpers::is_valid(validator.deref(), schema, item))
```
Here validator can be dereferenced before `iter()` so it won't be called multiple times (gave some improvement locally).
There are also other places in the PR where it can be applied

macisamuele replied:
Fair point , going to check for other similar occurrences.

macisamuele replied:
Addressed all the spotted occurrences in bbe3db6.

Stranger6667 commented:
I'll try to check it more thoroughly tomorrow

macisamuele commented:
@Stranger6667 👍  (rebasing on master in the meantime)

macisamuele commented:
I've rebased the branch on master and explicit implemented `Validate::is_valid` and `Validate::validate` as it was caused the performance drop that you were experiencing.

This was caused by the fact that the majority of the benchmarks are testing a single keyword and so they are definitely not benefitting from having the complete type dispatching.

I've still kept the type dispatching as it provides interesting performance improvements in realistic cases (multiple keywords to validate)

Updated raw results are available on https://gist.github.com/macisamuele/5f0a63e0e677f8d88d26540a6f0b10b5
**NOTE**: Benchmarks are covering the `is_valid` only API

98 benchmarks have performance improvements, while only 4 have regressed performances (still within ns of difference).
Validation on big-schema or small-schema present a performance improvement of 7.1% or better.

Stranger6667 commented:
@macisamuele 
It is absolutely great! A solid performance improvement! :) Thank you
There is only one thing that I'd like to have slightly differently - `unexpected` error, which I think we can eliminate completely with an affordable cost

macisamuele commented on src/error.rs:
```diff
@@ -151,6 +151,9 @@ pub enum ValidationErrorKind {
     UnknownReferenceScheme { scheme: String },
+
+    /// Unexpected error. This usually represent a bug into the validation
+    Unexpected { validator_representation: String },
```
The idea of having the `ValidationError::Unexpected` is to enable the definition of default `Validate::validate_*` methods which is very handy as it avoids a massive code duplication and keep binary size under control.

A possible alternative would be to use `panic!` or `unimplemented!` but it feels like "wrong" to abort the program for this reason.

So I originally felt that it was an acceptable middle-ground solution to have this "catch-all" validation error; understanding that it is the reflection of a bug it provides to the user such information and suggests to report it to the library such that can be fixed ASAP.

Please let me know what are your thoughts about this and if you need/want me to remove all the default implementations of `Validate::validate_`.

Stranger6667 replied:
Oh, I didn't see it initially, so I agree that this might be an acceptable middle-ground

Stranger6667 replied:
And I don't know how we can avoid this case statically :(

macisamuele replied:
I've tried re-iterating it again, looking for strategies to avoid this but nothing popped out to my eyes.

If you do agree we might have this merged and eventually opening an issue to try to remove this.
